### PR TITLE
EM27 webscraper

### LINKED ISSUE #109: Add webscraper for reading sensor values from EM27

### LINKED ISSUE #136: Initial delay in reading EM27 sensor values
The problem is that polling the server is done by a `QTimer`, but that currently means that the program waits for 2s before polling the server. Ideally we want to poll the server as soon as the program loads. This wouldn't be so much of a problem, but as the sensor values are added dynamically, it means that my window suddenly gets bigger 2s after loading, which is annoying.

@dc2917 You could probably fix this as part of #110.

### PR DESCRIPTION
**Pull request by dc2917**
This PR adds a module for obtaining the EM27 properties (the properties displayed in the "EM27 SOH Monitor" panel on the original GUI).

The properties are provided on a web server set up by the EM27's computer, thus this module provides web-scraping functionality to obtain the data from the corresponding web page. I have also included a snapshot of the webpage for testing functionality when access to the live server is not possible.

The module contains simple methods for opening and closing the connection to the server, reading the source HTML, and parsing this to return the data relating to the monitored properties. The expectation is that these data will be transmitted using `pubsub` to be displayed on the GUI.

Closes #109
Closes #136

### CODE REVIEW

**Comment by alexdewar** [comment]
Hey @dc2917. I'm guessing you forgot to tag me as a reviewer for this...?

**Comment by dc2917** [comment]
> Hey @dc2917. I'm guessing you forgot to tag me as a reviewer for this...?

I was going to give it one last look over before requesting reviews, but yeah I reckon it's review-ready now

**Review by alexdewar — CHANGES_REQUESTED** [review]
I've made some minor suggestions throughout, but tbh I think this needs a bit of a rewrite.

I'd start off by separating out the loading of the page and the parsing into separate places. It will allow you to test them separately and will also mean that your `DummyEM27Scraper` can just pass in the HTML from the file rather than a URL.

For loading the page, use the `requests` package rather than Python's built-in libraries; it's much nicer. (See [the other EM27 code](https://github.com/ImperialCollegeLondon/FINESSE/blob/main/finesse/hardware/em27_opus.py) if you want an example.) You should then be able to do all the loading of the page within a single function. The loading and the parsing don't actually need to keep track of any state, so they don't have to be class methods. You might even be able to get away with dropping the class altogether.

There are a couple of extra things I'd like you to add. Firstly, could you please add code to make it controllable with `pubsub`, because that's how we've been communicating with hardware from the frontend. I'd have one message to request values and one that responds with the values retrieved. (Again, the other EM27 code has an example of this.) Ideally, we will need another message signalling that an error has occurred, but if you get stuck with this then just submit it anyway and we can talk about it. Secondly, this needs some unit tests. Again, don't spend ages bashing your head against the wall if you can't get it to work -- ask if you need help!

Finally, I'd choose a catchier name for the files (blame Bruker for the naming). How about `em27_scraper.py` or `em27_sensors.py` or something?

**Review by dalonsoa — CHANGES_REQUESTED** [review]
I think this is a bit odd, with the webpage for the Dummy to load and commenting the `from ..config` line etc. I would remove the `if __init__` section, convert it into a test (a sort of integration or functional test) and leave the `em27_diag_autom` file as it would look in production. Otherwise, it looks rather half-baked.

And obviously, add the missing unit tests, but that can be incorporated as later stage. Open an issue about it.

File: finesse/hardware/em27_diag_autom.py
Comment (dc2917) [reply finesse/hardware/em27_diag_autom.py:18]: Yeah exactly, I think it belongs here rather than in the module containing the GUI component.

File: finesse/hardware/em27_diag_autom.py
Comment (dc2917) [reply finesse/hardware/em27_diag_autom.py:14]: The commented-out line doesn't work if I run this module as main (I get an `ImportError`).

Comment (alexdewar) [reply finesse/hardware/em27_diag_autom.py:14]: Does that happen even when you run it with `python -m finesse.hardware.em27_diag_autom`?

Comment (dc2917) [reply finesse/hardware/em27_diag_autom.py:14]: Ah, that works - thanks! Learnt something new

**Comment by alexdewar** [comment]
Btw you need to request a re-review from @dalonsoa because it won't let you merge this if one of the reviews is stuck at "changes requested".

File: tests/test_em27_scraper.py
Comment (alexdewar) [inline tests/test_em27_scraper.py:50]: You don't need to call `dev.read()` at all here, as a) it won't succeed and b) we're already testing it elsewhere. This test should just test `get_psf27sensor_data()`. There's no point in even checking `dev._is_read` because we know it will be `False` on any machine other than the one connected to the EM27.

The problem with this is that it only tests what happens when you pass an empty string into `get_psf27_sensor_data()`, which isn't particularly interesting! Define your own valid inputs for `get_psf27_sensor_data()` (and probably some invalid ones too) and check that the returned results are correct. Use a parametrised test for this.

File: tests/test_em27_scraper.py
Comment (alexdewar) [inline tests/test_em27_scraper.py:58]: Why are you silently ignoring all exceptions here?

File: tests/test_em27_scraper.py
Comment (alexdewar) [inline tests/test_em27_scraper.py:73]: This is good! But you don't actually need to raise the exception:

```suggestion
    dev._error_occurred(PSF27Error("Error occurred"))
    send_message_mock.assert_called_with("psf27.error", message="Error occurred")
```

File: tests/test_em27_scraper.py
Comment (alexdewar) [inline tests/test_em27_scraper.py:68]:
```suggestion
    """Test EM27Scraper's _error_occurred() method."""
```

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:32]: Why you need to ignore this line? Why is mypy complaining?

Not sure if related, but any tribute that starts with `_` should be treated as internal to the class and not be used externally. So, you should not use `self._poll_light._timer`. If this is a legitimate use of `_timer`, then it should not be "private" and the underscore should be removed.

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:97]: Same here.

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:94]: Maybe call this `set_data_table`. In any case, if this is all it is doing, why not making `data_table` public (ie. remove the underscore) and just assigned `data` to it in the calling function?

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:101]: And here.

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:105]: I missed this one - if it is going to be used outside of `poll_light`, `flash` also need to be public.

File: finesse/hardware/em27_scraper.py
Comment (dalonsoa) [inline finesse/hardware/em27_scraper.py:100]: This is a bit odd because there is only one exception to raise here and you know the error message, so what's the need to the `try... except`? Just do:
```suggestion
            if table_start == -1:
                self._error_occurred("PSF27Sensor table not found")
```
or:
```suggestion
            if table_start == -1:
                self._error_occurred(PSF27Error("PSF27Sensor table not found"))
```

File: tests/test_dummy_em27_scraper.py
Comment (dalonsoa) [reply tests/test_dummy_em27_scraper.py:30]: In general, `try... except` blocks do not make sense in tests. If you want to check if an exception occurred, use [`pytest.raises`](https://stackoverflow.com/a/29855337/3778792).

**Review by dalonsoa — CHANGES_REQUESTED** [review]
It's looking much better - well done! I've made a few comments of my own, but otherwise agreed with Alex's main comments.

File: finesse/hardware/em27_scraper.py
Comment (dc2917) [reply finesse/hardware/em27_scraper.py:47]: Honestly, no idea. Maybe point spread function if it's monitoring the optics? The table on the webpage is titled "EM27 SN_93 Automation units Diagnostics" which contains a load of information, some of which comes under sections whose titles start with "PSF27", so it seems like PSF27 is just a specific part of the EM27. The specific table I search for comes under "PSF27Sensor" so that's what I went with. Even more confusing is that the GUI panel which shows these is titled "EM27 SOH Monitor". Maybe I should ask Jon.

File: finesse/gui/interferometer_monitor.py
Comment (dc2917) [reply finesse/gui/interferometer_monitor.py:32]: The `_timer` attribute is a `QTimer`, and `mypy` and Qt don't seem to get along.

But yes, I've not been consistent with my public and private attributes as the code has evolved -I'll fix this.

File: finesse/gui/led_icons.py
Comment (alexdewar) [inline finesse/gui/led_icons.py:41]: Why is `timer` public?

File: finesse/hardware/dummy_em27_scraper.py
Comment (alexdewar) [inline finesse/hardware/dummy_em27_scraper.py:27]:
```suggestion
        except FileNotFoundError as e:
```

File: finesse/hardware/dummy_em27_scraper.py
Comment (alexdewar) [inline finesse/hardware/dummy_em27_scraper.py:28]:
```suggestion
            raise EM27Error(f"Dummy EM27 server file {self._url} not found") from e
```

File: finesse/hardware/em27_scraper.py
Comment (alexdewar) [inline finesse/hardware/em27_scraper.py:118]:
```suggestion
            logging.info("Read PSF27Sensor table")
            return request.text
        except (ConnectionError, HTTPError, Timeout) as e:
            raise EM27Error(f"Error connecting to {self._url}") from e
```

File: finesse/gui/led_icons.py
Comment (dalonsoa) [reply finesse/gui/led_icons.py:41]: Because it is used outside of the class. See my comments above.

**Review by alexdewar — APPROVED** [review]
I've made a couple of minor suggestions and there's one query, but otherwise I think this is good to go!

File: finesse/gui/led_icons.py
Comment (alexdewar) [reply finesse/gui/led_icons.py:41]: Ohhh I see.

File: finesse/gui/interferometer_monitor.py
Comment (dalonsoa) [inline finesse/gui/interferometer_monitor.py:103]: Maybe mention in the docstring that this function is needed because it is use via `pubsub`, otherwise, I would make the same comment as before - why you need a method that just set the data.

File: finesse/hardware/__init__.py
Comment (dalonsoa) [inline finesse/hardware/__init__.py:17]: As you're defining `scraper` below, you do not need to assign a type to it here.

File: finesse/hardware/dummy_em27_scraper.py
Comment (dalonsoa) [inline finesse/hardware/dummy_em27_scraper.py:13]: In our course of software engineering [we suggest that loading data in the constructor is a bad practice](https://imperialcollegelondon.github.io/grad_school_software_engineering_course/l1-05-abstractions/index.html) as there's a hidden dependency within the constructor that prevents initialising the class if not present - if should be done in a factory method. I don't have strong feelings - and I don't suggest to change this now as it will require significant updates in several places - but I just point this out for the future.

**Review by dalonsoa — APPROVED** [review]
I've made a couple of comments/suggestions, but otherwise is good to go.

**Pull request merged**
